docs/hacking/coding_best_practices.md

```diff
@@ -0,0 +1,280 @@
+   then it helps future reader if in a calling scope a variable that was to be
+   passed to `bar` was also called `foo`.
+
+ - Split your functions and methods. Helper functions defined either on module
+   level or even dynamically within a code object that uses them will make the
+   actual code so much easier to follow. Do this if:
+   - Your function is clearly doing several things in sequence (each one
+     deserves a helper);
+   - Your function is longer than about 30 lines;
+   - Your function has many levels of nesting;
+   - You feel the need to write a brief comment explaining what the next code
+     block is for;
+   A good function is a boring sequence of statements with little to no
+   branching which tells a story of simple transformations of arguments into
+   output.
+
+ - Avoid list comprehensions that span more than one line. This is usually a
+   sign to either fall back to a for loop or to add some intermediate
+   abstractions.
+
+ - Aliases are cheap and can make code look much simpler when properly named.
+
+ - Type aliases can make a huge difference for a reader:
+   ```
+   RawBarJSON = type(dict[str, Any])
+   def foo(bars: list[RawBarJSON])
+   ```
+   vs
+   ```
+   def foo(bars: list[dict[str, Any]])
+   ```
+   Please use them. They also act as a form of documentation.
+
+ - Strive for declarative code: what the code does is often of greater interest
+   than how it does it.  Having a rich set of aliases and helper functions
+   greatly helps with this.
+
+ - Try to avoid reinventing the wheel. Python has a rich collection of built-in
+   components with well-known behavior, which allows one to write concise and
+   idiomatic code. It usually pays off to check if `collections`, `itertools` or
+   `functools` contain something that could be reused.
+
+ - When possible and practical implement
+   [dunder methods](https://docs.python.org/3/reference/datamodel.html#basic-customization)
+   for you classes.
+
+ - When defining a base class consider making it abstract, this would
+   disambiguate the intent. While there is some room for doubt with
+   `NotImplementedError` (maybe it would be just fine if I have implemented it
+   right here?) there is no doubt with `@abc.abstractmethod`.
+
+### Favor styles which are easier to reason about
+
+ - When possible prefer immutable objects, especially when defining containers
+   with constants.  Use `frozenset` when there are no plans to extend it, use
+   tuples instead of lists for static data.
+
+ - When possible prefer building new, modified container objects from
+   old ones to modifying existing container objects in place. A function that
+   takes an argument and constructs a return value basing on it is easier to
+   reason about than a function that modifies its argument.
+
+ - Refresh recommendations from  [PEP8](https://peps.python.org/pep-0008)
+   and [PEP20](https://peps.python.org/pep-0020);
+
+
+### Write clear tests which actually test functionality
```

Comment:
Should we drop the[ test guidelines](https://github.com/containerbuildsystem/cachi2/blob/main/CONTRIBUTING.md#test-guidelines) from CONTRIBUTING.md?

Reply:
No. That guide is the main one, this is supplementary.